source/common/upstream/upstream_impl.cc

```diff
@@ -177,7 +177,7 @@ void HostSetImpl::updateHosts(HostVectorConstSharedPtr hosts,
   // level WRR works, we would age out the existing entries via picks and lazily
   // apply the new weights.
   if (hosts_per_locality_ != nullptr && locality_weights_ != nullptr &&
-      !locality_weights_->empty()) {
+      !locality_weights_->empty() && !healthy_hosts_->empty()) {
```

Comment:
I haven't fully tracked how this portion of the code works, but don't we still want to have a locality with no healthy hosts if we end up in panic mode? (Don't remember where we landed on panic mode in this path.) @htuch ?

Reply:
https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/load_balancer_impl.cc#L355-L368

if i'm reading this right we check for panic mode before we attempt to pick a locality, so it shouldn't matter

Reply:
Technically though, couldn't we not be in global panic, but still not have any healthy hosts in an individual locality? This code has gone through a bunch of revisions and I'm not as familiar with it as I once was. We should wait until @htuch has time to weigh in.

Reply:
I looked through the code a bit, mostly to increase my understanding, and I guess in the case where global panic % doesn't cover the case in which a single selected locality has any healthy hosts, we would move on to the next try and take all healthy hosts, which I guess makes sense (basically abort locality routing). Though, IMO, this brings up a pretty important point which is that I think we really need to have a per-locality panic threshold, since you could wind up in a situation where when doing locality routing you are hammering a very small number of locality hosts without doing any spillover. I think this is worth some more detailed discussion/thinking before we commit to a solution (unless I am missing something).

Reply:
From what I understand the `healthy_hosts_` vector is the list of healthy hosts in the *priority*, not locality. I think we'll only run into this if no priorities have any healthy hosts - otherwise we'd have selected a priority with healthy hosts. This means that the only case this PR modifies is the case when all priorities are unhealthy and panic mode is disabled: if panic mode was enabled we'd enter the `isGlobalPanic` and never attempt to select a locality. 

re per locality: based on https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/load_balancing.html#locality-weighted-load-balancing isn't there already spillover for localities? or are you talking about the case where a locality goes completely unhealthy and other localities are unable to handle the load? 

in any case I don't think this PR affects the case where a single locality within a priority is unhealthy, unless i'm misunderstanding what `healthy_hosts_` means.

Reply:
So, yeah, we have spillover across both localities and priorities based on health. I think if all but a few hosts in a locality were unhealthy, we would already be spilling to the next priority level. 

As far as this PR goes, I think this is a legitimate fix in expected use. The reasoning is that `healthy_hosts_` being non-empty implies that there must exist at least one locality where there is a non-zero weight. That is, assuming each host belongs to a locality, from a reading of `eds.cc` this seems to be the case.

I think if you are doing priority/locality spillover, you only want to panic once you are right at the bottom of the barrel, or is there some case that doesn't hold?

Reply:
Was just looking at this when you commented. Yes, I think this makes sense. With spillover, I think looking at panic threshold for the entire priority level should be sufficient for preventing cascading failover due to a single locality having a small number of healthy hosts (if someone configures it that way). @snowp do you mind adding some more comments? This code is very complicated and between locality routing and zone aware routing it can be hard to figure out the flow without a really detailed reading.

Reply:
Added a comment. Let me know if anything is inaccurate or if I should add more information